src/wp-includes/class-wp-script-modules.php

```diff
@@ -363,4 +366,119 @@ private function get_src( string $id ): string {
 
 		return $src;
 	}
+
+	/**
+	 * Print data associated with Script Modules.
+	 *
+	 * The data will be embedded in the page HTML and can be read by Script Modules on page load.
+	 *
+	 * @since 6.6.0
+	 *
+	 * Data can be associated with a Script Module via the
+	 * {@see "scriptmoduledata_{$module_id}"} filter.
+	 *
+	 * The data for a Script Module will be serialized as JSON in a script tag with an ID of the
+	 * form `wp-scriptmodule-data_{$module_id}`.
+	 */
+	public function print_script_module_data(): void {
+		$modules = array();
+		foreach ( array_keys( $this->get_marked_for_enqueue() ) as $id ) {
+			$modules[ $id ] = true;
+		}
+		foreach ( array_keys( $this->get_import_map()['imports'] ) as $id ) {
+			$modules[ $id ] = true;
+		}
+
+		foreach ( array_keys( $modules ) as $module_id ) {
+			/**
+			 * Filters data associated with a given Script Module.
+			 *
+			 * Script Modules may require data that is required for initialization or is essential
+			 * to have immediately available on page load. These are suitable use cases for
+			 * this data.
+			 *
+			 * The dynamic portion of the hook name, `$module_id`, refers to the Script Module ID
+			 * that the data is associated with.
+			 *
+			 * This is best suited to pass essential data that must be available to the module for
+			 * initialization or immediately on page load. It does not replace the REST API or
+			 * fetching data from the client.
+			 *
+			 * @example
+			 *   add_filter(
+			 *     'scriptmoduledata_MyScriptModuleID',
+			 *     function ( array $data ): array {
+			 *       $data['script-needs-this-data'] = 'ok';
+			 *       return $data;
+			 *     }
+			 *   );
+			 *
+			 * If the filter returns no data (an empty array), nothing will be embedded in the page.
+			 *
+			 * The data for a given Script Module, if provided, will be JSON serialized in a script
+			 * tag with an ID of the form `wp-scriptmodule-data_{$module_id}`.
+			 *
+			 * The data can be read on the client with a pattern like this:
+			 *
+			 * @example
+			 *   const dataContainer = document.getElementById( 'wp-scriptmodule-data_MyScriptModuleID' );
+			 *   let data = {};
+			 *   if ( dataContainer ) {
+			 *     try {
+			 *       data = JSON.parse( dataContainer.textContent );
+			 *     } catch {}
+			 *   }
+			 *   initMyScriptModuleWithData( data );
+			 *
+			 * @since 6.6.0
+			 *
+			 * @param array $data The data that should be associated with the array.
```

Comment:
Why must the data be an array? Couldn't it be any JSON-serializable data?  What if `null` is passed instead as the default value, and any value other than `null` will get a JSON `script` printed?

Reply:
This question came up on the Gutenberg PR here: https://github.com/WordPress/gutenberg/pull/61658#discussion_r1620502804

This should be an array because it makes the interface for filters clear. If modules start to use other types then the filter interface becomes less clear. They receive and return an array: `filter_func( array $data): array`.

It's true that this data is tightly coupled to the module and its needs, so if it really just needs a string then why not let it pass that? I still think it's likely a good practice because as that module evolves and requires more or different data it doesn't need to change types and start dealing with an object instead of a string. This could also allow versioned data to be passed or for different versions of modules to access only the data they need.